processing/src/main/java/io/druid/segment/StringDimensionIndexer.java

```diff
@@ -534,79 +535,6 @@ public void fillBitmapsFromUnsortedEncodedArray(
     }
   }
 
-  @Override
-  public ValueMatcher makeIndexingValueMatcher(
-      final String matchValue,
-      final IncrementalIndexStorageAdapter.EntryHolder holder,
-      final int dimIndex
-  )
-  {
-    final int encodedVal = getEncodedValue(matchValue, false);
-    final boolean matchOnNull = Strings.isNullOrEmpty(matchValue);
-    if (encodedVal < 0 && !matchOnNull) {
-      return new BooleanValueMatcher(false);
-    }
-
-    return new ValueMatcher()
```

Comment:
My concern is that the new realtime dimension value matcher (`StringValueMatcherColumnSelectorStrategy`, line 51) is less specialised than this one. The difference between benchmarks and production is that benchmarks run the same code (same impls of some interfaces, called in hot loops) again and again, so the calls are monomorphic and inlined, but in production they stay virtual, because different queries cause different impls of some interfaces appear in hot loops. In this particular case, the version of ValueMatcher from `StringValueMatcherColumnSelectorStrategy` may specialize `getRow()` because it is monomorphic, then eliminate IndexedInts allocation because it can prove that it doesn't escape, and then essentially run the same code as here, in `StringDimensionIndexer. makeIndexingValueMatcher()` but in production `getRow()` in `StringValueMatcherColumnSelectorStrategy` won't be inlined, so `IndexedInts` indirection above the `int[]` array will remain.

So I suggest to add `DimensionSelector.makeValueMatcher()` anyway, maybe as a separate PR, but necessarily before in the same Druid release as this PR.